File: src/ch/CH.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

class CH
{
    public static List<int> getLevelNodes(Graph graph, short level)
    {
        var set = new List<int>();
        int node_count = graph.nodeCount();
        for (int i=0; i < node_count; i++)
        {
            if (graph.getNodeLevel(i) >= level)
            {
                set.Add(i);
            }
        }
        return set;
    }

    public static List<int> getShortestPath(Graph graph, int start, int end, short level=-1)
    {
        var edges = new List<int>();
        var heap = new PriorityQueue<int, int>();
        heap.Enqueue(start, 0);
        var flags = new Flag[graph.nodeCount()];
        for (int i = 0; i < flags.Length; i++)
        {
            flags[i].pathlength = 1000000000;
        }
        flags[start].pathlength = 0;
        int curr_id = -1;

        while (true)
        {
            try
            {
                curr_id = heap.Dequeue();
            }
            catch (Exception)
            {
                return edges;
            }
            if (curr_id == end)
            {
                break;
            }
            Node curr = graph.getNode(curr_id);
            ref Flag currflag = ref flags[curr_id];
            if (currflag.visited)
            {
                continue;
            }
            currflag.visited = true;
            foreach (int e_id in curr.edges)
            {
                Edge edge = graph.getEdge(e_id);
                int other_id = graph.getOtherNode(e_id, curr_id);
                Node other = graph.getNode(other_id);
                ref Flag otherflag = ref flags[other_id];
                if (otherflag.visited || other.level < level)
                {
                    continue;
                }
                int newlength = currflag.pathlength + edge.weight;
                if (otherflag.pathlength > newlength)
                {
                    otherflag.prevEdge = e_id;
                    otherflag.pathlength = newlength;
                    heap.Enqueue(other_id, newlength);
                }
            }
        }

        curr_id = end;
        int edge_id;
        while (true)
        {
            if (curr_id == start)
            {
                break;
            }
            edge_id = flags[curr_id].prevEdge;
            edges.Add(edge_id);
            curr_id = graph.getOtherNode(edge_id, curr_id);
        }
        edges.Reverse();
        return edges;
    }

    public static List<int> getShortestPathCH(Graph graph, int start, int end, short level=-1)
    {
        var edges = new List<int>();
        var start_heap = new PriorityQueue<int, int>();
        start_heap.Enqueue(start, 0);
        var end_heap = new PriorityQueue<int, int>();
        end_heap.Enqueue(end, 0);
        var flags = new FlagCH[graph.nodeCount()];
        for (int i = 0; i < flags.Length; i++)
        {
            flags[i].pathlength1 = 1000000000;
            flags[i].pathlength2 = 1000000000;
        }
        flags[start].pathlength1 = 0;
        flags[end].pathlength2 = 0;
        int curr_id = -1;
        int mid_id = -1;

        while (true)
        {
            // from start
            try
            {
                curr_id = start_heap.Dequeue();
            }
            catch (Exception)
            {
                return edges;
            }
            Node curr = graph.getNode(curr_id);
            ref FlagCH currflag = ref flags[curr_id];
            if (currflag.visited1)
            {
                continue;
            }
            if (currflag.visited2)
            {
                mid_id = curr_id;
                break;
            }
            currflag.visited1 = true;
            foreach (int e_id in curr.edges)
            {
                Edge edge = graph.getEdge(e_id);
                int other_id = graph.getOtherNode(e_id, curr_id);
                Node other = graph.getNode(other_id);
                ref FlagCH otherflag = ref flags[other_id];
                if (otherflag.visited1 || other.level <= curr.level)
                {
                    continue;
                }
                int newlength = currflag.pathlength1 + edge.weight;
                if (otherflag.pathlength1 > newlength)
                {
                    otherflag.prevEdge1 = e_id;
                    otherflag.pathlength1 = newlength;
                    start_heap.Enqueue(other_id, newlength);
                }
            }

            // from end
            try
            {
                curr_id = end_heap.Dequeue();
            }
            catch (Exception)
            {
                return edges;
            }
            curr = graph.getNode(curr_id);
            currflag = ref flags[curr_id];
            if (currflag.visited2)
            {
                continue;
            }
            if (currflag.visited1)
            {
                mid_id = curr_id;
                break;
            }
            currflag.visited2 = true;
            foreach (int e_id in curr.edges)
            {
                Edge edge = graph.getEdge(e_id);
                int other_id = graph.getOtherNode(e_id, curr_id);
                Node other = graph.getNode(other_id);
                ref FlagCH otherflag = ref flags[other_id];
                if (otherflag.visited2 || other.level <= curr.level)
                {
                    continue;
                }
                int newlength = currflag.pathlength2 + edge.weight;
                if (otherflag.pathlength2 > newlength)
                {
                    otherflag.prevEdge2 = e_id;
                    otherflag.pathlength2 = newlength;
                    end_heap.Enqueue(other_id, newlength);
                }
            }
        }


        curr_id = mid_id;
        int edge_id;
        while (true)
        {
            if (curr_id == start)
            {
                break;
            }
            edge_id = flags[curr_id].prevEdge1;
            var e = new List<int>();
            getEdgesFromShortcut(e, graph, edge_id);
            edges.AddRange(e);
            curr_id = graph.getOtherNode(edge_id, curr_id);
        }
        edges.Reverse();
        curr_id = mid_id;
        while (true)
        {
            if (curr_id == end)
            {
                break;
            }
            edge_id = flags[curr_id].prevEdge2;
            var e = new List<int>();
            getEdgesFromShortcut(e, graph, edge_id);
            edges.AddRange(e);
            curr_id = graph.getOtherNode(edge_id, curr_id);
        }
        return edges;
    }

    public static List<int> getShortestPathCH2(Graph graph, int start, int end, short level=-1)
    {
        var edges = new List<int>();
        var heap = new PriorityQueue<(int, bool), int>();
        heap.Enqueue((start, true), 0);
        heap.Enqueue((end, false), 0);
        var flags = new FlagCH[graph.nodeCount()];
        for (int i = 0; i < flags.Length; i++)
        {
            flags[i].pathlength1 = 1000000000;
            flags[i].pathlength2 = 1000000000;
        }
        flags[start].pathlength1 = 0;
        flags[end].pathlength2 = 0;
        int curr_id;
        int mid_id = -1;
        bool from_start;

        while (true)
        {
            // from start
            try
            {
                (curr_id, from_start) = heap.Dequeue();
            }
            catch (Exception)
            {
                return edges;
            }
            Node curr = graph.getNode(curr_id);
            ref FlagCH currflag = ref flags[curr_id];
            if ((currflag.visited2 && from_start) || (currflag.visited1 && !from_start))
            {
                mid_id = curr_id;
                break;
            }
            if (from_start)
                currflag.visited1 = true;
            else
                currflag.visited2 = true;
            foreach (int e_id in curr.edges)
            {
                Edge edge = graph.getEdge(e_id);
                int other_id = graph.getOtherNode(e_id, curr_id);
                Node other = graph.getNode(other_id);
                ref FlagCH otherflag = ref flags[other_id];
                if ((otherflag.visited1 && from_start) || (otherflag.visited2 && !from_start))
                if (other.level <= curr.level)
                {
                    continue;
                }
                if (from_start) {
                    int newlength = currflag.pathlength1 + edge.weight;
                    if (otherflag.pathlength1 > newlength)
                    {
                        otherflag.prevEdge1 = e_id;
                        otherflag.pathlength1 = newlength;
                        heap.Enqueue((other_id, true), newlength);
                    }
                }
                else {
                    int newlength = currflag.pathlength1 + edge.weight;
                    if (otherflag.pathlength2 > newlength)
                    {
                        otherflag.prevEdge2 = e_id;
                        otherflag.pathlength2 = newlength;
                        heap.Enqueue((other_id, false), newlength);
                    }
                }
            }
        }
        curr_id = mid_id;
        int edge_id;
        while (true)
        {
            if (curr_id == start)
            {
                break;
            }
            edge_id = flags[curr_id].prevEdge1;
            var e = new List<int>();
            getEdgesFromShortcut(e, graph, edge_id);
            edges.AddRange(e);
            curr_id = graph.getOtherNode(edge_id, curr_id);
        }
        edges.Reverse();
        curr_id = mid_id;
        while (true)
        {
            if (curr_id == end)
            {
                break;
            }
            edge_id = flags[curr_id].prevEdge2;
            var e = new List<int>();
            getEdgesFromShortcut(e, graph, edge_id);
            edges.AddRange(e);
            curr_id = graph.getOtherNode(edge_id, curr_id);
        }
        return edges;
    }

    public static void getEdgesFromShortcut(List<int> edges, Graph graph, int edge_id)
    {
        var edge = graph.getEdge(edge_id);
        if (edge.is_shortcut)
        {
            var shortcut = graph.GetShortcut(edge.shortcut_id);
            foreach (int e in shortcut.edges)
            {
                getEdgesFromShortcut(edges, graph, e);
            }
        }
        else
        {
            edges.Add(edge_id);
        }
    }

    public static void calcContraction(Graph graph)
    {
        Console.WriteLine("started contracting graph");
        for (int i=0; i<graph.nodeCount(); i++)
        {
            graph.setNodeLevel(i, 0);
        }
        short level = 0;
        var nodes = getLevelNodes(graph, level);
        while (true)
        {
            int nc1 = nodes.Count;
            if (nodes.Count == 0)
            {
                break;
            }
            Console.WriteLine("start ordering nodes");
            nodes = orderNodes(graph, nodes, level);
            Console.WriteLine("finished ordering nodes");
            int ec1 = graph.edgeCount();
            contractLevel2(graph, nodes, level);
            int ec2 = graph.edgeCount();
            nodes = getLevelNodes(graph, (short)(level+1));
            int nc2 = nodes.Count;
            Console.WriteLine($"contracted level {level}: {ec2-ec1} shortcuts added, {nc1-nc2}/{nc1} nodes contracted");
            ec1 = ec2;
            level += 1;
        }
        Console.WriteLine("finished contracting graph");
    }

    public static void contractLevel(Graph graph, List<int> nodes, short level)
    {
        short new_level = (short)(level + 1);

        var rand = new Random();
        while (nodes.Count > 0)
        {
            int node_id = nodes[rand.Next(0, nodes.Count)];
            nodes.Remove(node_id);
            Node node = graph.getNode(node_id);
            var neigbours = new List<int>();
            foreach (int edge_id in node.edges)
            {
                Edge edge = graph.getEdge(edge_id);
                int other_id = graph.getOtherNode(edge_id, node_id);
                Node other = graph.getNode(other_id);
                if (other.level < level)
                {
                    continue;
                }
                neigbours.Add(other_id);
            }
            foreach (int n in neigbours)
            {
                nodes.Remove(n);
                graph.setNodeLevel(n, new_level);
            }
            
            for (int i=0; i<neigbours.Count; i++)
            {
                for (int j=i+1; j<neigbours.Count; j++)
                {
                    int from = neigbours[i];
                    int to = neigbours[j];
                    var edges = getShortestPath(graph, from, to, level);
                    bool add_shortcut = false;
                    edges.ForEach((id) => {
                        Edge e = graph.getEdge(id);
                        if (e.node_b == node_id || e.node_a == node_id) {
                            add_shortcut = true;
                        }
                    });
                    if (add_shortcut == false) {
                        continue;
                    }
                    int weight = 0;
                    foreach (int e in edges) {
                        weight += graph.getWeight(e);
                    }
                    graph.addEdge(from, to, weight, true, edges);
                }
            }
        }
    }

    public static List<int> orderNodes(Graph graph, List<int> nodes, short level)
    {
        int[] counts = new int[nodes.Count];
        var mapping = new Dictionary<int, int>();
        for (int i=0; i<nodes.Count; i++)
        {
            mapping[nodes[i]] = i;
        }

        var rand = new Random(DateTime.Now.Millisecond);
        int n = 1;
        if (nodes.Count > 50000)
            n = 25;
        else if (nodes.Count > 20000)
            n = 10;
        else if (nodes.Count > 6000)
            n = 3;
        for (int i=0; i<nodes.Count/n; i++)
        {
            int start = nodes[rand.Next(0, nodes.Count)];
            int end = nodes[rand.Next(0, nodes.Count)];
            var edges = getShortestPath(graph, start, end, level);
            foreach (int e_id in edges)
            {
                Edge edge = graph.getEdge(e_id);
                counts[mapping[edge.node_a]] += 1;
                counts[mapping[edge.node_b]] += 1;
            }
        }

        var out_list = nodes.OrderBy((i) => {
            return counts[mapping[i]];
        }).ToList();
        return out_list;
    }

    public static void contractLevel2(Graph graph, List<int> nodes, short level)
    {
        short new_level = (short)(level + 1);

        while (nodes.Count > 0)
        {
            int node_id = nodes[0];
            nodes.Remove(node_id);
            Node node = graph.getNode(node_id);
            var neigbours = new List<int>();
            foreach (int edge_id in node.edges)
            {
                Edge edge = graph.getEdge(edge_id);
                int other_id = graph.getOtherNode(edge_id, node_id);
                Node other = graph.getNode(other_id);
                if (other.level < level)
                {
                    continue;
                }
                neigbours.Add(other_id);
            }
            foreach (int n in neigbours)
            {
                nodes.Remove(n);
                graph.setNodeLevel(n, new_level);
            }
            
            for (int i=0; i<neigbours.Count; i++)
            {
                for (int j=i+1; j<neigbours.Count; j++)
                {
                    int from = neigbours[i];
                    int to = neigbours[j];
                    var edges = getShortestPath(graph, from, to, level);
                    bool add_shortcut = false;
                    edges.ForEach((id) => {
                        Edge e = graph.getEdge(id);
                        if (e.node_b == node_id || e.node_a == node_id) {
                            add_shortcut = true;
                        }
                    });
                    if (add_shortcut == false) {
                        continue;
                    }
                    int weight = 0;
                    foreach (int e in edges) {
                        weight += graph.getWeight(e);
                    }
                    graph.addEdge(from, to, weight, true, edges);
                }
            }
        }
    }

    public static Graph loadGraph(string url)
    {
        Graph graph = new Graph();

        FileInfo f = new FileInfo(url);
        if (!f.Exists || f.Name.Split(".")[1] != "graph")
        {
            throw new FileNotFoundException("specified path doesnt meet requirements");
        }
        string filename = f.Name.Split(".")[0];
        Byte[] graphdata = File.ReadAllBytes(url);
        MemoryStream graphstream = new MemoryStream(graphdata);
        BinaryReader graphreader = new BinaryReader(graphstream);
        int nodecount = graphreader.ReadInt32();
        int edgecount = graphreader.ReadInt32();
        int startindex = 8 + nodecount * 5 + edgecount * 8;
        MemoryStream edgerefstream = new MemoryStream(graphdata, startindex, graphdata.Length - startindex);
        BinaryReader edgerefreader = new BinaryReader(edgerefstream);
        for (int i = 0; i < nodecount; i++)
        {
            int s = graphreader.ReadInt32();
            sbyte c = graphreader.ReadSByte();
            for (int j = 0; j < c; j++)
            {
                edgerefreader.ReadInt32();
            }
            graph.addNode(-1);
        }
        for (int i = 0; i < edgecount; i++)
        {
            int start = graphreader.ReadInt32();
            int end = graphreader.ReadInt32();
            graph.addEdge(start, end, -1);
        }
        graphreader.Close();
        edgerefreader.Close();
        graphstream.Close();
        edgerefstream.Close();

        Byte[] weightdata = File.ReadAllBytes(f.DirectoryName + "/" + filename + "-weight");
        MemoryStream weightstream = new MemoryStream(weightdata);
        BinaryReader weightreader = new BinaryReader(weightstream);
        for (int i = 0; i < edgecount; i++)
        {
            graph.setWeight(i, weightreader.ReadByte());
        }
        weightreader.Close();
        weightstream.Close();
        
        return graph;
    }
}

struct Flag
{
    public int pathlength;
    public int prevEdge;
    public bool visited;
}

struct FlagCH
{
    public int pathlength1;
    public int prevEdge1;
    public bool visited1;

    public int pathlength2;
    public int prevEdge2;
    public bool visited2;
}

File: src/go-routing/routing/ch.go
package routing

import (
	"fmt"

	"github.com/ttpr0/simple-routing-visualizer/src/go-routing/geo"
	"github.com/ttpr0/simple-routing-visualizer/src/go-routing/graph"
	. "github.com/ttpr0/simple-routing-visualizer/src/go-routing/util"
)

type flag_ch struct {
	path_length1 float64
	path_length2 float64
	prev_edge1   int32
	is_shortcut1 bool
	prev_edge2   int32
	visited1     bool
	visited2     bool
	is_shortcut2 bool
}

type CH struct {
	startheap   PriorityQueue[int32, float64]
	endheap     PriorityQueue[int32, float64]
	mid_id      int32
	start_id    int32
	end_id      int32
	path_length float64
	graph       graph.ICHGraph
	flags       Dict[int32, flag_ch]
}

func NewCH(graph graph.ICHGraph, start, end int32) *CH {
	startheap := NewPriorityQueue[int32, float64](10)
	endheap := NewPriorityQueue[int32, float64](10)
	flags := NewDict[int32, flag_ch](100)

	flags[start] = flag_ch{path_length1: 0, visited1: false, prev_edge1: -1, is_shortcut1: false, path_length2: 1000000, visited2: false, prev_edge2: -1, is_shortcut2: false}
	startheap.Enqueue(start, 0)
	flags[end] = flag_ch{path_length1: 1000000, visited1: false, prev_edge1: -1, is_shortcut1: false, path_length2: 0, visited2: false, prev_edge2: -1, is_shortcut2: false}
	endheap.Enqueue(end, 0)

	ch := CH{
		startheap:   startheap,
		endheap:     endheap,
		mid_id:      -1,
		start_id:    start,
		end_id:      end,
		path_length: 100000000,
		graph:       graph,
		flags:       flags,
	}

	return &ch
}

func (self *CH) CalcShortestPath() bool {
	explorer := self.graph.GetDefaultExplorer()

	for {
		if self.startheap.Len() == 0 && self.endheap.Len() == 0 {
			break
		}
		if self.mid_id != -1 {
			s_id, _ := self.startheap.Peek()
			e_id, _ := self.endheap.Peek()
			s_flag := self.flags[s_id]
			e_flag := self.flags[e_id]
			if s_flag.path_length1 >= self.path_length && e_flag.path_length2 >= self.path_length {
				break
			}
		}

		// from start
		if self.startheap.Len() != 0 {
			curr_id, _ := self.startheap.Dequeue()
			curr_flag := self.flags[curr_id]
			if curr_flag.visited1 {
				continue
			}
			curr_flag.visited1 = true
			if curr_flag.visited2 && self.path_length > (curr_flag.path_length1+curr_flag.path_length2) {
				self.mid_id = curr_id
				self.path_length = curr_flag.path_length1 + curr_flag.path_length2
			}
			explorer.ForAdjacentEdges(curr_id, graph.FORWARD, graph.ADJACENT_ALL, func(ref graph.EdgeRef) {
				edge_id := ref.EdgeID
				other_id := ref.OtherID
				if self.graph.GetNodeLevel(other_id) <= self.graph.GetNodeLevel(curr_id) {
					return
				}
				var other_flag flag_ch
				if self.flags.ContainsKey(other_id) {
					other_flag = self.flags[other_id]
				} else {
					other_flag = flag_ch{path_length1: 1000000, visited1: false, prev_edge1: -1, is_shortcut1: false, path_length2: 1000000, visited2: false, prev_edge2: -1, is_shortcut2: false}
				}
				weight := explorer.GetEdgeWeight(ref)
				new_length := curr_flag.path_length1 + float64(weight)
				if new_length < other_flag.path_length1 {
					other_flag.path_length1 = new_length
					other_flag.prev_edge1 = edge_id
					other_flag.is_shortcut1 = ref.IsShortcut()
					self.startheap.Enqueue(other_id, new_length)
				}
				self.flags[other_id] = other_flag
			})
			self.flags[curr_id] = curr_flag
		}

		// from end
		if self.endheap.Len() != 0 {
			curr_id, _ := self.endheap.Dequeue()
			curr_flag := self.flags[curr_id]
			if curr_flag.visited2 {
				continue
			}
			curr_flag.visited2 = true
			if curr_flag.visited1 && self.path_length > (curr_flag.path_length1+curr_flag.path_length2) {
				self.mid_id = curr_id
				self.path_length = curr_flag.path_length1 + curr_flag.path_length2
			}
			explorer.ForAdjacentEdges(curr_id, graph.BACKWARD, graph.ADJACENT_ALL, func(ref graph.EdgeRef) {
				edge_id := ref.EdgeID
				other_id := ref.OtherID
				if self.graph.GetNodeLevel(other_id) <= self.graph.GetNodeLevel(curr_id) {
					return
				}
				var other_flag flag_ch
				if self.flags.ContainsKey(other_id) {
					other_flag = self.flags[other_id]
				} else {
					other_flag = flag_ch{path_length1: 1000000, visited1: false, prev_edge1: -1, is_shortcut1: false, path_length2: 1000000, visited2: false, prev_edge2: -1, is_shortcut2: false}
				}
				weight := explorer.GetEdgeWeight(ref)
				new_length := curr_flag.path_length2 + float64(weight)
				if new_length < other_flag.path_length2 {
					other_flag.path_length2 = new_length
					other_flag.prev_edge2 = edge_id
					other_flag.is_shortcut2 = ref.IsShortcut()
					self.endheap.Enqueue(other_id, new_length)
				}
				self.flags[other_id] = other_flag
			})
			self.flags[curr_id] = curr_flag
		}
	}
	if self.mid_id == -1 {
		return false
	}
	return true
}

func (self *CH) Steps(count int, visitededges *List[geo.CoordArray]) bool {
	edges := NewList[int32](100)
	defer func() {
		fmt.Println("running")
		for _, edge := range edges {
			visitededges.Add(self.graph.GetEdgeGeom(edge))
		}
	}()

	explorer := self.graph.GetDefaultExplorer()
	for c := 0; c < count; c++ {
		if self.startheap.Len() == 0 && self.endheap.Len() == 0 {
			return false
		}
		if self.mid_id != -1 {
			s_id, _ := self.startheap.Peek()
			e_id, _ := self.endheap.Peek()
			s_flag := self.flags[s_id]
			e_flag := self.flags[e_id]
			if s_flag.path_length1 >= self.path_length && e_flag.path_length2 >= self.path_length {
				return false
			}
		}
		// from start
		if self.startheap.Len() != 0 {
			curr_id, _ := self.startheap.Dequeue()
			curr_flag := self.flags[curr_id]
			if curr_flag.visited1 {
				continue
			}
			curr_flag.visited1 = true
			if curr_flag.visited2 && self.path_length > (curr_flag.path_length1+curr_flag.path_length2) {
				self.mid_id = curr_id
				self.path_length = curr_flag.path_length1 + curr_flag.path_length2
			}
			explorer.ForAdjacentEdges(curr_id, graph.FORWARD, graph.ADJACENT_ALL, func(ref graph.EdgeRef) {
				edge_id := ref.EdgeID
				other_id := ref.OtherID
				if self.graph.GetNodeLevel(other_id) <= self.graph.GetNodeLevel(curr_id) {
					return
				}
				var other_flag flag_ch
				if self.flags.ContainsKey(other_id) {
					other_flag = self.flags[other_id]
				} else {
					other_flag = flag_ch{path_length1: 1000000, visited1: false, prev_edge1: -1, is_shortcut1: false, path_length2: 1000000, visited2: false, prev_edge2: -1, is_shortcut2: false}
				}
				weight := explorer.GetEdgeWeight(ref)

				if ref.IsShortcut() {
					self.graph.GetEdgesFromShortcut(&edges, edge_id, false)
				} else {
					edges.Add(edge_id)
				}

				new_length := curr_flag.path_length1 + float64(weight)
				if new_length < other_flag.path_length1 {
					other_flag.path_length1 = new_length
					other_flag.prev_edge1 = edge_id
					other_flag.is_shortcut1 = ref.IsShortcut()
					self.startheap.Enqueue(other_id, new_length)
				}
				self.flags[other_id] = other_flag
			})
			self.flags[curr_id] = curr_flag
		}

		if self.mid_id != -1 {
			break
		}
		// from end
		if self.endheap.Len() != 0 {
			curr_id, _ := self.endheap.Dequeue()
			curr_flag := self.flags[curr_id]
			if curr_flag.visited2 {
				continue
			}
			curr_flag.visited2 = true
			if curr_flag.visited1 && self.path_length > (curr_flag.path_length1+curr_flag.path_length2) {
				self.mid_id = curr_id
				self.path_length = curr_flag.path_length1 + curr_flag.path_length2
			}
			explorer.ForAdjacentEdges(curr_id, graph.BACKWARD, graph.ADJACENT_ALL, func(ref graph.EdgeRef) {
				edge_id := ref.EdgeID
				other_id := ref.OtherID
				if self.graph.GetNodeLevel(other_id) <= self.graph.GetNodeLevel(curr_id) {
					return
				}
				var other_flag flag_ch
				if self.flags.ContainsKey(other_id) {
					other_flag = self.flags[other_id]
				} else {
					other_flag = flag_ch{path_length1: 1000000, visited1: false, prev_edge1: -1, is_shortcut1: false, path_length2: 1000000, visited2: false, prev_edge2: -1, is_shortcut2: false}
				}
				weight := explorer.GetEdgeWeight(ref)
				if ref.IsShortcut() {
					self.graph.GetEdgesFromShortcut(&edges, edge_id, false)
				} else {
					edges.Add(edge_id)
				}
				new_length := curr_flag.path_length2 + float64(weight)
				if new_length < other_flag.path_length2 {
					other_flag.path_length2 = new_length
					other_flag.prev_edge2 = edge_id
					other_flag.is_shortcut2 = ref.IsShortcut()
					self.endheap.Enqueue(other_id, new_length)
				}
				self.flags[other_id] = other_flag
			})
			self.flags[curr_id] = curr_flag
		}
	}
	return true
}

func (self *CH) GetShortestPath() Path {
	explorer := self.graph.GetDefaultExplorer()

	path := NewList[int32](10)
	length := int32(self.flags[self.mid_id].path_length1 + self.flags[self.mid_id].path_length2)
	curr_id := self.mid_id
	for {
		if curr_id == self.start_id {
			break
		}
		curr_flag := self.flags[curr_id]
		if curr_flag.is_shortcut1 {
			self.graph.GetEdgesFromShortcut(&path, curr_flag.prev_edge1, true)
			curr_id = explorer.GetOtherNode(graph.CreateCHShortcutRef(curr_flag.prev_edge1), curr_id)
		} else {
			path.Add(curr_flag.prev_edge1)
			curr_id = explorer.GetOtherNode(graph.CreateEdgeRef(curr_flag.prev_edge1), curr_id)
		}
	}
	for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
		path[i], path[j] = path[j], path[i]
	}
	curr_id = self.mid_id
	for {
		if curr_id == self.end_id {
			break
		}
		curr_flag := self.flags[curr_id]
		if curr_flag.is_shortcut2 {
			self.graph.GetEdgesFromShortcut(&path, curr_flag.prev_edge2, false)
			curr_id = explorer.GetOtherNode(graph.CreateCHShortcutRef(curr_flag.prev_edge2), curr_id)
		} else {
			path.Add(curr_flag.prev_edge2)
			curr_id = explorer.GetOtherNode(graph.CreateEdgeRef(curr_flag.prev_edge2), curr_id)
		}
	}
	fmt.Println("length:", length)
	return NewPath(self.graph, path)
}
